### ISSUE TITLE
If the prompt fills the entire line, clr_eol confuses some terminals

### ISSUE THREAD

**Issue by faho** [fish-shell/fish-shell#8164]
@ridiculousfish in #8002:

> Ah ha, we can cheat by just clearing the line first, before outputting its contents. Easy fix!

Yeah, no, that cheat doesn't work. It breaks repainting when the line got shorter - #8163.

**Comment by mqudsi** [comment]
I think clearing first and outputting second (what you just reverted) would work if we used `\e[2K` instead of `clr_eol` (I don't know what curses calls it). That would clear the entire line regardless of the cursor position. At least, if I understand what you're saying correctly.

**Comment by faho** [comment]
Either that, or moving the cursor to the beginning before clearing the line - `\r` and then clr_eol.

I can't find the "clear entire line" sequence in the terminfo man page, only clr_bol and clr_eol, so we probably can't use it (oh how I hate the terminfo system).

**Comment by ajhall** [comment]
https://github.com/IlanCosman/tide/issues/310

This bug is easy to trigger with one of the built-in 2-line prompts in https://github.com/IlanCosman/tide while using Windows Terminal, Alacritty, foot, WezTerm, etc.

It doesn't break anything essential, though, it's just a minor visual annoyance.

**Comment by akuntsch** [comment]
This also happens in Termux on a Galaxy Tab S7+ for me (also using the tide prompt).

**Comment by davidmartos96** [comment]
@faho Is this an easy fix or is there something advanced involved for the right solution?
From the `tide` linked issue, this is an simple prompt that reproduces the issue on terminals such as alacritty or wezterm.
The printed dashed line is missing the last character.
```fish
function fish_prompt
    string repeat --count=$COLUMNS -
    echo $PWD
end
```

Unfortunately it's an annoying bug when the clipped character clips useful information, such a version number or a git hash.

I tried implementing the proposed solutions from above, but I'm not able to avoid https://github.com/fish-shell/fish-shell/issues/8163 from reappearing (I don't have much experience with `term.h` and CSI sequences, so it could very well be a faulty implementation on my part)

**Comment by faho** [comment]
>@faho Is this an easy fix or is there something advanced involved for the right solution?

@davidmartos96 All I know is what I wrote above - it should be possible to reset the cursor to the beginning of the line at the right time so we clear the entire line. But I haven't implemented it so I don't know for sure.

**Comment by gagbo** [comment]
Would it be possible to solve specifically the "long prompt" problem by conditionnally clearing to EOL?

```diff
diff --git a/src/screen.cpp b/src/screen.cpp
index c22834e5d..f47b63501 100644
--- a/src/screen.cpp
+++ b/src/screen.cpp
@@ -763,7 +763,8 @@ void screen_t::update(const wcstring &left_prompt, const wcstring &right_prompt,
         size_t start = 0;
         for (const size_t line_break : left_prompt_layout.line_breaks) {
             this->write_str(left_prompt.substr(start, line_break - start));
-            if (clr_eol) {
+            // Clear the rest of the line if the printed line was shorter that the screen width
+            if (clr_eol && (line_break - start) < (this->actual.screen_width - 1)) {
                 this->write_mbs(clr_eol);
             }
             start = line_break;
```

The reason the first tentative fix was reverted was that "never emitting `clr_eol`" was also wrong for the cases where the prompt to overwrite was longer than the new version. This patch (if the `line_break - start` is actually matching the count of columns) would catch it by clearing the rest of the line if there are cells left on the line.

This would still be buggy for the specific case where the new prompt is _exactly_ 1 character short of a full line: when the cursor is on the last character of the line, I don't think there's a way to check whether we just repainted over it, or if we need to clear the cell.

**Comment by diniamo** [comment]
I want to add 2 things:

According to foot's developer, this behavior is a part of the VT specification, meaning the fix has to be made in fish:

> This is according to the VT specification. You'll find that e.g XTerm behaves the same way.

\- https://codeberg.org/dnkl/foot/issues/1804#issuecomment-2170363

<br>

The sequence mentioned above *is* a part of terminfo, and it's called `delete_line`. I have successfully patched the 3.7.1 release with this: [fish-prompt-fix.patch](https://github.com/diniamo/niqspkgs/blob/721afaee98ffdd58c992e38116f26454d95ae905/pkgs/patches/fish-prompt-fix.patch) (however I ultimately went with the patch in the comment above, because that can avoid clearing altogether most of the time).

**Comment by budimanjojo** [comment]
This issue is really annoying. Afaik `zellij`, `tmux`, and `kitty`are the terminal/multiplexer that's not affected by this. Is the patch above acceptable to get merged into fish? @faho @ridiculousfish

**Comment by diniamo** [comment]
The `delete_line` one should be, but it has to be ported to Rust.

**Comment by budimanjojo** [comment]
I'm looking at the current source code, don't know `rust` but according to this comment: https://github.com/fish-shell/fish-shell/blob/48a65506883a76aeea9f8268d039fb630ba82d8d/src/screen.rs#L942-L946
This bug shouldn't be happening with the rust build?

**Comment by diniamo** [comment]
Well, were you trying on the Rust build? The latest release is still C.

**Comment by gagbo** [comment]
> I mean, if this bug is not gonna affect the future Rust build then there's no reason to create a PR to fix this?

@diniamo’s point is that if you didn’t test the Rust build to assess that the bug is gone, we should let this issue opened, that’s all.

If you’ve built the Rust rewrite of fish on `main`/`master` locally, and could not reproduce this bug, then we can close the issue with a "Won’t fix (in C++)" or "Fixed (in next release)" tag

**Comment by diniamo** [comment]
> I mean, if this bug is not gonna affect the future Rust build then there's no reason to create a PR to fix this?

Correct, it's fine to leave the issue open as long as the latest release has the bug.

> I still see this bug using foot terminal on the latest main/master build (3.7.1-2327-g48a655068)

Well that's interesting, what's with that comment then?

> hopefully it's as simple as `pkgs.buildRustPackage` on `nix`.

It's far from that, I started porting the derivation to Rust, but I either ran into some stupid issue that I had no clue how to fix, or concluded that it was too much effort, I don't remember.

**Comment by diniamo** [comment]
And so I did: https://github.com/fish-shell/fish-shell/blob/master/Cargo.toml#L27, fish has a git dependency, I never figured out how to make those work.

**Comment by budimanjojo** [comment]
@diniamo I managed to get fish to build with something like this:
```nix
{
  rustPlatorm,
  fetchFromGitHub,
  cmake,
}:

rustPlatform.buildRustPackage {
  pname = "fish";
  version = "1";

  src = fetchFromGitHub {
    owner = "fish-shell";
    repo = "fish-shell";
    rev = "master";
    hash = "sha256-MAXa8FrErG/B71A0kC0rVNUZxJEVWg/ucfUpjow5Weo=";
  };

  cargoLock = {
    # just copy the lock file to the local path and source it here
    lockFile = ./Cargo.lock;
    allowBuiltinFetchGit = true;
  };

  checkFlags = [
    # these tests are failing on me
    "--skip=builtins::tests::test_tests::test_test_builtin"
    "--skip=tests::highlight::test_highlighting"
    "--skip=tests::highlight::test_is_potential_path"
  ];

  nativeBuildInputs = [ cmake ];
```

But, even after building the package, I got other error when trying to run the binary:
```
 result/bin/fish
error: Fish cannot find its asset files in '/usr/local/share/fish'. Refusing to read configuration.
budiman@budimanjojo-main /home/budiman/Github/nix-config > exit
```
I think it's because we also need to have `postPatch` and `postInstall` similar to the C++ build in `nixpkgs` and I don't know how to :cry:

**Comment by diniamo** [comment]
That's far from correct unfortunately.

**Comment by budimanjojo** [comment]
What's not correct? My guess is because `buildRustPackage` only put the binary files in `$out/bin/` and not `$out/share/`. And we also need to tell `fish` to find them in `$out/share` instead of `/usr/local/share/fish/`.

**Comment by diniamo** [comment]
Fish still uses cmake, which buildRustWhatever doesn't work with. You also need to patch the assets.

**Comment by budimanjojo** [comment]
Yeah building Nix package is too much for my brain. Maybe I can try building it in a Ubuntu VM when I have the time.

**Comment by budimanjojo** [comment]
There's a nightly builds PPA for Ubuntu too https://github.com/fish-shell/fish-shell/wiki/Development-builds#ubuntu-ppa-instructions

I'll do this tomorrow

**Comment by budimanjojo** [comment]
Very weird, I can't reproduce this issue in my Ubuntu VM even with the current stable version. Is there something else triggering this problem? Tried in Gnome Terminal and Alacritty.

**Comment by krobelus** [comment]
you don't need to build a package to run a recent version.
`cd fish-shell && make && build/fish` should be enough.
(It's broken in some irrelevant ways (completions) but those can be fixed if we can manage to convince people that this should not be broken. To work around the breakage you can `cd build && cmake -DCMAKE_INSTALL_PREFIX=/path/to/installation && make install`)

**Comment by budimanjojo** [comment]
Okay I managed to reproduce this with the nightly build from `ppa:fish-shell/nightly-master` in Ubuntu VM using `contour` terminal (weird that this is not happening in `alacritty` and `gnome-terminal`).
Fish version is `3.7.1-2327`

I also tried running it like the step by @krobelus above and the issue is there too.
![image](https://github.com/user-attachments/assets/8913f65e-64e7-4575-b45b-ab95c79ebb5f)

**Comment by budimanjojo** [comment]
Thank you @krobelus! I'm trying it out right now

**Comment by budimanjojo** [comment]
Oops, now `contour` works correctly with the latest commit, but `alacritty` is now broken with the same issue :D

**Comment by diniamo** [comment]
@budimanjojo I've managed to do it: https://github.com/diniamo/niqspkgs/blob/main/pkgs/fish-git.nix (flake ref `github:diniamo/niqspkgs#fish-git`.

While from some basic testing it seems to work, I might have missed some stuff, especially when it comes to patching the command overrides to use absolute paths (though this would never actually cause an issue in practice).

**Comment by budimanjojo** [comment]
Thank you!

**Comment by jnns** [comment]
Although I'm on Fish 4.0.2, I still experience the missing rightmost character on the prompt in Ghostty. The weird thing is that the last character is shown when I'm running `fish` inside the current shell process. 

As you can see in the screenshot, the right prompt is missing a character when I exit the sub-process using Ctrl-D: 
<img width="788" height="114" alt="Image" src="https://github.com/user-attachments/assets/e762298d-9c3f-4e9a-9a65-0a2f3c1ca5c5" />

**Comment by krobelus** [comment]
> Although I'm on Fish 4.0.2, I still experience the missing rightmost character on the prompt in Ghostty. The weird thing is that the last character is shown when I'm running `fish` inside the current shell process. 
> 
> As you can see in the screenshot, the right prompt is missing a character when I exit the sub-process using Ctrl-D: 
> <img width="788" height="114" alt="Image" src="https://github.com/user-attachments/assets/e762298d-9c3f-4e9a-9a65-0a2f3c1ca5c5" />

@jnns can you give steps to reproduce, preferably in a new issue?
E.g. something like

1. configure fish with

	function fish_prompt
	    string repeat $COLUMNS -
	    echo "$PWD> "
	end
	function fish_right_prompt
	    echo asdf
	end

2. launch fish
3. inside fish launch another fish
4. press ctrl-d

these example steps work fine for me

**Comment by jnns** [comment]
Thank you for your assistance @krobelus. Your example worked flawlessly which made me assume that it was a problem with my Tide prompt config. Upon inspection, I found out that I seem to be running with two different Fish shell versions and that the problem only occurs in version 3.

```
❯ tide bug-report
[...snip...]

Please copy the following information into the issue:

fish version: 3.7.0
tide version: 6.1.1
[...snip...]
❯ fish --version
fish, version 4.0.2
```

I'll see how to fix the issue on my system. Thanks again!

UPDATE: The problem was indeed that I had /usr/local/bin/fish (ver 3) and /opt/homebrew/bin/fish (ver 4). Possibly through some weird laptop migration issue a few months ago. I removed the old Fish installation manually and now it works.
